### ISSUE TITLE
How to Reduce Variance in a Final Machine Learning Model

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#437]
How to Reduce Variance in a Final Machine Learning Model<br>
<br>
https://ift.tt/2MEOV4E<br>
<br>
Jason Brownlee<br>
<br>
<p>A final machine learning model is one trained on all available data and is then used to make predictions on new data.</p>
<p>A problem with most final models is that they suffer variance in their predictions.</p>
<p>This means that each time you fit a model, you get a slightly different set of parameters that in turn will make slightly different predictions. Sometimes more and sometimes less skillful than what you expected.</p>
<p>This can be frustrating, especially when you are looking to deploy a model into an operational environment.</p>
<p>In this post, you will discover how to think about model variance in a final model and techniques that you can use to reduce the variance in predictions from a final model.</p>
<p>After reading this post, you will know:</p>
<ul>
<li>The problem with variance in the predictions made by a final model.</li>
<li>How to measure model variance and how variance is addressed generally when estimating parameters.</li>
<li>Techniques you can use to reduce the variance in predictions made by a final model.</li>
</ul>
<p>Let&rsquo;s get started.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/08/How-to-Reduce-Variance-in-a-Final-Machine-Learning-Model.jpg" alt="How to Reduce Variance in a Final Machine Learning Model" width="640" class="wp-image-5985" height="426">
<p class="wp-caption-text">How to Reduce Variance in a Final Machine Learning Model<br>
Photo by <a href="https://www.flickr.com/photos/kirt_edblom/41951791832/">Kirt Edblom</a>, some rights reserved.</p>
</div>
<h2>Final Model</h2>
<p>Once you have discovered which model and model hyperparameters result in the best skill on your dataset, you&rsquo;re ready to prepare a final model.</p>
<p>A final model is trained on all available data, e.g. the training and the test sets.</p>
<p>It is the model that you will use to make predictions on new data were you do not know the outcome.</p>
<p>The final model is the outcome of your applied machine learning project.</p>
<p>To learn more about preparing a final model, see the post:</p>
<ul>
<li><a href="https://machinelearningmastery.com/train-final-machine-learning-model/">How to Train a Final Machine Learning Model</a></li>
</ul>
<h2>Bias and Variance</h2>
<p>The bias-variance trade-off is a conceptual idea in applied machine learning to help understand the sources of error in models.</p>
<ul>
<li>
<strong>Bias</strong> refers to assumptions in the learning algorithm that narrow the scope of what can be learned. This is useful as it can accelerate learning and lead to stable results, at the cost of the assumption differing from reality.</li>
<li>
<strong>Variance</strong> refers to the sensitivity of the learning algorithm to the specifics of the training data, e.g. the noise and specific observations. This is good as the model will be specialized to the data at the cost of learning random noise and varying each time it is trained on different data.</li>
</ul>
<p>The bias-variance tradeoff is a conceptual tool to think about these sources of error and how they are always kept in balance.</p>
<p>More bias in an algorithm means that there is less variance, and the reverse is also true.</p>
<p>You can learn more about the bias-variance tradeoff in this post:</p>
<ul>
<li><a href="https://machinelearningmastery.com/gentle-introduction-to-the-bias-variance-trade-off-in-machine-learning/">Gentle Introduction to the Bias-Variance Trade-Off in Machine Learning</a></li>
</ul>
<p>You can control this balance.</p>
<p>Many machine learning algorithms have hyperparameters that directly or indirectly allow you to control the bias-variance tradeoff.</p>
<p>For example, the <em>k</em> in <em>k</em>-nearest neighbors is one example. A small <em>k</em> results in predictions with high variance and low bias. A large <em>k</em> results in predictions with a small variance and a large bias.</p>
<h2>The Problem of Variance in Final Models</h2>
<p>Most final models have a problem: they suffer from variance.</p>
<p>Each time a model is trained by an algorithm with high variance, you will get a slightly different result.</p>
<p>The slightly different model in turn will make slightly different predictions, for better or worse.</p>
<p>This is a problem with training a final model as we are required to use the model to make predictions on real data where we do not know the answer and we want those predictions to as good as possible.</p>
<p>We want to the best possible version of the model that we can get.</p>
<p>We want the variance to play out in our favor.</p>
<p>If we can&rsquo;t achieve that, at least we want the variance to not fall against us when making predictions.</p>
<h2>Measure Variance in the Final Model</h2>
<p>There are two common sources of variance in a final model:</p>
<ul>
<li>The noise in the training data.</li>
<li>The use of randomness in the machine learning algorithm.</li>
</ul>
<p>The first type we introduced above.</p>
<p>The second type impacts those algorithms that harness randomness during learning.</p>
<p>Three common examples include:</p>
<ul>
<li>Choice of random split points in random forest.</li>
<li>Random weight initialization in neural networks.</li>
<li>Shuffling training data in stochastic gradient descent.</li>
</ul>
<p>You can measure both types of variance in your specific model using your training data.</p>
<ul>
<li>
<strong>Measure Algorithm Variance</strong>: The variance introduced by the stochastic nature of the algorithm can be measured by repeating the evaluation of the algorithm on the same training dataset and calculating the variance or standard deviation of the model skill.</li>
<li>
<strong>Measure Training Data Variance</strong>: The variance introduced by the training data can be measured by repeating the evaluation of the algorithm on different samples of training data, but keeping the seed for the pseudorandom number generator fixed then calculating the variance or standard deviation of the model skill.</li>
</ul>
<p>Often, the combined variance is estimated by running repeated k-fold cross-validation on a training dataset then calculating the variance or standard deviation of the model skill.</p>
<h2>Reduce Variance of an Estimate</h2>
<p>If we want to reduce the amount of variance in a prediction, we must add bias.</p>
<p>Consider the case of a simple statistical estimate of a population parameter, such as estimating the mean from a small random sample of data.</p>
<p>A single estimate of the mean will have high variance and low bias.</p>
<p>This is intuitive because if we repeated this process 30 times and calculated the standard deviation of the estimated mean values, we would see a large spread.</p>
<p>The solutions for reducing the variance are also intuitive.</p>
<p>Repeat the estimate on many different small samples of data from the domain and calculate the mean of the estimates, leaning on the <a href="https://machinelearningmastery.com/a-gentle-introduction-to-the-central-limit-theorem-for-machine-learning/">central limit theorem</a>.</p>
<p>The mean of the estimated means will have a lower variance. We have increased the bias by assuming that the average of the estimates will be a more accurate estimate than a single estimate.</p>
<p>Another approach would be to dramatically increase the size of the data sample on which we estimate the population mean, leaning on the <a href="https://machinelearningmastery.com/a-gentle-introduction-to-the-law-of-large-numbers-in-machine-learning/">law of large numbers</a>.</p>
<h2>Reduce Variance of a Final Model</h2>
<p>The principles used to reduce the variance for a population statistic can also be used to reduce the variance of a final model.</p>
<p>We must add bias.</p>
<p>Depending on the specific form of the final model (e.g. tree, weights, etc.) you can get creative with this idea.</p>
<p>Below are three approaches that you may want to try.</p>
<p>If possible, I recommend designing a test harness to experiment and discover an approach that works best or makes the most sense for your specific data set and machine learning algorithm.</p>
<h3>1. Ensemble Predictions from Final Models</h3>
<p>Instead of fitting a single final model, you can fit multiple final models.</p>
<p>Together, the group of final models may be used as an ensemble.</p>
<p>For a given input, each model in the ensemble makes a prediction and the final output prediction is taken as the average of the predictions of the models.</p>
<p>A sensitivity analysis can be used to measure the impact of ensemble size on prediction variance.</p>
<h3>2. Ensemble Parameters from Final Models</h3>
<p>As above, multiple final models can be created instead of a single final model.</p>
<p>Instead of calculating the mean of the predictions from the final models, a single final model can be constructed as an ensemble of the <a href="https://machinelearningmastery.com/difference-between-a-parameter-and-a-hyperparameter/">parameters</a> of the group of final models.</p>
<p>This would only make sense in cases where each model has the same number of parameters, such as neural network weights or regression coefficients.</p>
<p>For example, consider a linear regression model with three coefficients [b0, b1, b2]. We could fit a group of linear regression models and calculate a final b0 as the average of b0 parameters in each model, and repeat this process for b1 and b2.</p>
<p>Again, a sensitivity analysis can be used to measure the impact of ensemble size on prediction variance.</p>
<h3>3. Increase Training Dataset Size</h3>
<p>Leaning on the law of large numbers, perhaps the simplest approach to reduce the model variance is to fit the model on more training data.</p>
<p>In those cases where more data is not readily available, perhaps data augmentation methods can be used instead.</p>
<p>A sensitivity analysis of training dataset size to prediction variance is recommended to find the point of diminishing returns.</p>
<h2>Fragile Thinking</h2>
<p>There are approaches to preparing a final model that aim to get the variance in the final model to work for you rather than against you.</p>
<p>The commonality in these approaches is that they seek a single best final model.</p>
<p>Two examples include:</p>
<ul>
<li>
<strong>Why not fix the random seed?</strong> You could fix the random seed when fitting the final model. This will constrain the variance introduced by the stochastic nature of the algorithm.</li>
<li>
<strong>Why not use early stopping?</strong> You could check the skill of the model against a holdout set during training and stop training when the skill of the model on the hold set starts to degrade.</li>
</ul>
<p>I would argue that these approaches and others like them are fragile.</p>
<p>Perhaps you can gamble and aim for the variance to play-out in your favor. This might be a good approach for machine learning competitions where there is no real downside to losing the gamble.</p>
<p>I won&rsquo;t.</p>
<p>I think it&rsquo;s safer to aim for the best average performance and limit the downside.</p>
<p>I think that the trick with navigating the bias-variance tradeoff for a final model is to think in samples, not in terms of single models. To optimize for average model performance.</p>
<h2>Further Reading</h2>
<p>This section provides more resources on the topic if you are looking to go deeper.</p>
<ul>
<li><a href="https://machinelearningmastery.com/train-final-machine-learning-model/">How to Train a Final Machine Learning Model</a></li>
<li><a href="https://machinelearningmastery.com/gentle-introduction-to-the-bias-variance-trade-off-in-machine-learning/">Gentle Introduction to the Bias-Variance Trade-Off in Machine Learning</a></li>
<li><a href="https://en.wikipedia.org/wiki/Bias%E2%80%93variance_tradeoff">Bias-Variance Tradeoff on Wikipedia</a></li>
<li>
<a href="https://arxiv.org/abs/1710.03282">Checkpoint Ensembles: Ensemble Methods from a Single Training Process</a>, 2017.</li>
</ul>
<h2>Summary</h2>
<p>In this post, you discovered how to think about model variance in a final model and techniques that you can use to reduce the variance in predictions from a final model.</p>
<p>Specifically, you learned:</p>
<ul>
<li>The problem with variance in the predictions made by a final model.</li>
<li>How to measure model variance and how variance is addressed generally when estimating parameters.</li>
<li>Techniques you can use to reduce the variance in predictions made by a final model.</li>
</ul>
<p>Do you have any questions?<br>
Ask your questions in the comments below and I will do my best to answer.</p>
<p>The post <a rel="nofollow" href="https://machinelearningmastery.com/how-to-reduce-model-variance/">How to Reduce Variance in a Final Machine Learning Model</a> appeared first on <a rel="nofollow" href="https://machinelearningmastery.com">Machine Learning Mastery</a>.</p>
<br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via Machine Learning Mastery https://ift.tt/2if3vo8<br>
<br>
August 13, 2018 at 03:10AM
